File: Java/dependency-injection/changing-behavior-through-composition/src/main/java/com/github/grzesiek_galezowski/ChangingBehaviorThroughComposition/Alarms/SwitchableAlarm.cs
namespace ChangingBehaviorThroughComposition
{
  public class SwitchableAlarm : Alarm
  {
    private readonly Alarm _defaultAlarm;
    private readonly Alarm _secondaryAlarm;
    private readonly SwitchCriteria _switchCriteria;

    public SwitchableAlarm(SwitchCriteria switchCriteria, Alarm defaultAlarm, Alarm secondaryAlarm)
    {
      _defaultAlarm = defaultAlarm;
      _secondaryAlarm = secondaryAlarm;
      _switchCriteria = switchCriteria;
    }
  
    public void Trigger()
    {
      if(_switchCriteria.IsNotMet())
      {
        _defaultAlarm.Trigger();
      }
      else
      {
        _secondaryAlarm.Trigger();
      }
    }

    public void Disable()
    {
      _defaultAlarm.Disable();
      _secondaryAlarm.Disable();
    }
  }
}

File: Java/dependency-injection/changing-behavior-through-composition/src/main/java/com/github/grzesiek_galezowski/ChangingBehaviorThroughComposition/Alarms/SwitchableAlarm.java
package com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.Alarms;

import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.Alarm;
import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.SwitchCriteria;

public class SwitchableAlarm implements Alarm {
  private final Alarm _defaultAlarm;
  private final Alarm _secondaryAlarm;
  private final SwitchCriteria _switchCriteria;

  public SwitchableAlarm(SwitchCriteria switchCriteria, Alarm defaultAlarm, Alarm secondaryAlarm) {
    _defaultAlarm = defaultAlarm;
    _secondaryAlarm = secondaryAlarm;
    _switchCriteria = switchCriteria;
  }

  public void trigger() {
    if (_switchCriteria.isNotMet()) {
      _defaultAlarm.trigger();
    } else {
      _secondaryAlarm.trigger();
    }
  }

  public void disable() {
    _defaultAlarm.disable();
    _secondaryAlarm.disable();
  }
}
